src/main/java/com/devcycle/sdk/server/local/bucketing/LocalBucketing.java

```diff
@@ -208,32 +221,71 @@ public FlushPayload[] flushEventQueue(String token) throws JsonProcessingExcepti
         return payloads;
     }
 
-    public void onPayloadFailure(String token, String payloadId, boolean retryable) {
-        int tokenAddress = newWasmString(token);
+    public void onPayloadFailure(String sdkKey, String payloadId, boolean retryable) {
+        unpinAll();
+        int sdkKeyAddress = getSDKKeyAddress(sdkKey);
         int payloadIdAddress = newWasmString(payloadId);
 
         Func onPayloadFailurePtr = linker.get(store, "", "onPayloadFailure").get().func();
         WasmFunctions.Consumer3<Integer, Integer, Integer> fn = WasmFunctions.consumer(store, onPayloadFailurePtr, I32, I32, I32);
-        fn.accept(tokenAddress, payloadIdAddress, retryable ? 1 : 0);
+        fn.accept(sdkKeyAddress, payloadIdAddress, retryable ? 1 : 0);
     }
 
-    public void onPayloadSuccess(String token, String payloadId) {
-        int tokenAddress = newWasmString(token);
+    public void onPayloadSuccess(String sdkKey, String payloadId) {
+        unpinAll();
+        int sdkKeyAddress = getSDKKeyAddress(sdkKey);
         int payloadIdAddress = newWasmString(payloadId);
 
         Func onPayloadSuccessPtr = linker.get(store, "", "onPayloadSuccess").get().func();
         WasmFunctions.Consumer2<Integer, Integer> fn = WasmFunctions.consumer(store, onPayloadSuccessPtr, I32, I32);
-        fn.accept(tokenAddress, payloadIdAddress);
+        fn.accept(sdkKeyAddress, payloadIdAddress);
     }
 
-    public int getEventQueueSize(String token) {
-        int tokenAddress = newWasmString(token);
+    public int getEventQueueSize(String sdkKey) {
+        unpinAll();
+        int sdkKeyAddress = getSDKKeyAddress(sdkKey);
 
         Func getEventQueueSizePtr = linker.get(store, "", "eventQueueSize").get().func();
         WasmFunctions.Function1<Integer, Integer> getEventQueueSize = WasmFunctions.func(
                 store, getEventQueueSizePtr, I32, I32);
 
-        return getEventQueueSize.call(tokenAddress);
+        return getEventQueueSize.call(sdkKeyAddress);
+    }
+
+    private void pinParameter(int address) {
+        Func pinPtr = linker.get(store, "", "__pin").get().func();
+        WasmFunctions.Consumer1<Integer> pin = WasmFunctions.consumer(store, pinPtr, I32);
+        pin.accept(address);
+    }
+
+    private void unpinParameter(int address) {
+        Func unpinPtr = linker.get(store, "", "__unpin").get().func();
+        WasmFunctions.Consumer1<Integer> unpin = WasmFunctions.consumer(store, unpinPtr, I32);
+        unpin.accept(address);
+    }
+
+    private void unpinAll() {
+        for(int address : pinnedAddresses) {
+            unpinParameter(address);
+        }
+        pinnedAddresses.clear();
+    }
+
+    private int getPinnedParameter(String param) {
+        int address = newWasmString(param);
+        pinParameter(address);
+        pinnedAddresses.add(address);
```

Comment:
this should be happening inside `pinParameter`

Reply:
actually sorry no it shouldnt because the SDK key shouldnt be in here

Reply:
ah right, because the SDK key addresses are separate.. yup